activemodel/lib/active_model/dirty.rb

```diff
@@ -252,6 +252,12 @@ def attribute_changed_in_place?(attr_name) # :nodoc:
       mutations_from_database.changed_in_place?(attr_name)
     end
 
+    def rewind_attribute_assignments # :nodoc
+      if defined?(@attributes) && mutations_before_last_save.respond_to?(:attributes)
```

Comment:
Why do we need to check if AttributeMutationTracker responds to attributes?

Reply:
[`NullTransaction`](https://github.com/rails/rails/blob/548bd509857ca5bba300915aa5ad9bff6e8c0238/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L84)s don't respond to `#attributes`, and having it return an empty attribute set isn't what we want either. I can refactor to match the type via `#is_a?` instead, but this covers more cases.

Reply:
This defeats a bit the purpose of the null object pattern, but I understand why we can't return an empty attributes array

Reply:
What I could do instead, is forgo reaching into the mutation tracker and store the previous attributes in an ivar. Matching the type also goes against the null object pattern.